plugins/parsers/avro/parser.go

```diff
@@ -0,0 +1,220 @@
+	}
+
+	if len(metrics) != 1 {
+		return nil, errors.New("line contains multiple metrics")
+	}
+
+	return metrics[0], nil
+}
+
+func (p *Parser) SetDefaultTags(tags map[string]string) {
+	p.DefaultTags = tags
+}
+
+func (p *Parser) createMetric(data map[string]interface{}, schema string) (telegraf.Metric, error) {
+	// Tags differ from fields, in that tags are inherently strings.
+	// fields can be of any type.
+	fields := make(map[string]interface{})
+	tags := make(map[string]string)
+
+	// Set default tag values
+	for k, v := range p.DefaultTags {
+		tags[k] = v
+	}
+	// Avro doesn't have a Tag/Field distinction, so we have to tell
+	// Telegraf which items are our tags.
+	for _, tag := range p.Tags {
+		sTag, err := internal.ToString(data[tag])
+		if err != nil {
+			p.Log.Warnf("Could not convert %v to string for tag %q: %v", data[tag], tag, err)
+		}
+		tags[tag] = sTag
+	}
+	var fieldList []string
+	if len(p.Fields) != 0 {
+		// If you have specified your fields in the config, you
+		// get what you asked for.
+		fieldList = p.Fields
+	} else {
+		for k := range data {
+			// Otherwise, that which is not a tag is a field
+			if _, ok := tags[k]; !ok {
+				fieldList = append(fieldList, k)
+			}
+		}
+	}
+	// We need to flatten out our fields.  The default (the separator
+	// string is empty) is equivalent to what streamreactor does.
+	sep := flatten.SeparatorStyle{
+		Before: "",
+		Middle: p.FieldSeparator,
+		After:  "",
+	}
+	for _, fld := range fieldList {
+		candidate := make(map[string]interface{})
+		candidate[fld] = data[fld] // 1-item map
+		flat, err := flatten.Flatten(candidate, "", sep)
+		if err != nil {
+			return nil, fmt.Errorf("flatten field %q failed: %w", fld, err)
+		}
+		for k, v := range flat {
+			fields[k] = v
+		}
+	}
+
+	var schemaObj map[string]interface{}
+	if err := json.Unmarshal([]byte(schema), &schemaObj); err != nil {
+		return nil, fmt.Errorf("unmarshaling schema failed: %w", err)
+	}
+	if len(fields) == 0 {
+		// A telegraf metric needs at least one field.
+		return nil, errors.New("number of fields is 0; unable to create metric")
+	}
+	// Now some fancy stuff to extract the measurement.
+	// If it's set in the configuration, use that.
+	name := p.Measurement
+	separator := "."
+	if name == "" {
+		// Try using the namespace defined in the schema. In case there
+		// is none, just use the schema's name definition.
+		nsStr, ok := schemaObj["namespace"].(string)
+		// namespace is optional
+		if !ok {
+			separator = ""
+		}
+
+		nStr, ok := schemaObj["name"].(string)
+		if !ok {
+			return nil, fmt.Errorf("could not determine name from schema %s", schema)
+		}
+		name = nsStr + separator + nStr
+	}
+	// Still don't have a name?  Guess we should use the metric name if
+	// it's set.
+	if name == "" {
+		name = p.MetricName
+	}
+	// Nothing?  Give up.
+	if name == "" {
+		return nil, errors.New("could not determine measurement name")
+	}
+	return metric.New(name, tags, fields, time.Now()), nil
```

Comment:
Isn't this always setting the time to now and never using the user's specified timestamp value?